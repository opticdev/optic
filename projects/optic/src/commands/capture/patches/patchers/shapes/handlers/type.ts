import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import {
  JsonSchemaKnownKeyword,
  ErrorObject,
  ShapeDiffResult,
  ShapeDiffResultKind,
} from '../diff';
import { Schema, SchemaObject } from '../schema';
import { ShapeLocation } from '../documented-bodies';
import { SupportedOpenAPIVersions } from '@useoptic/openapi-io';
import { ShapePatch } from '../patches';
import { CapturedInteraction } from '../../../../sources/captured-interactions';
import { OperationGroup, PatchImpact } from '../../spec/patches';

export function* typeKeywordDiffs(
  validationError: ErrorObject,
  example: any
): IterableIterator<ShapeDiffResult> {
  if (validationError.keyword !== JsonSchemaKnownKeyword.type) return;
  const typeKeywordPath = jsonPointerHelpers.decode(
    validationError.schemaPath.substring(1)
  );

  const propertyPath = jsonPointerHelpers.pop(
    jsonPointerHelpers.compile(typeKeywordPath)
  );

  const keyName = jsonPointerHelpers.decode(propertyPath).pop() || '';

  const unmatchedValue = jsonPointerHelpers.get(
    example,
    validationError.instancePath
  );

  yield {
    description: `'${keyName}' did not match schema`,
    expectedType: `${validationError.params.type}`,
    kind: ShapeDiffResultKind.UnmatchedType,
    keyword: JsonSchemaKnownKeyword.type,
    instancePath: validationError.instancePath,
    propertyPath: propertyPath,
    key: keyName,
    example: unmatchedValue,
  };
}
export function* typePatches(
  diff: ShapeDiffResult,
  schema: SchemaObject,
  interaction: CapturedInteraction,
  shapeContext: { location?: ShapeLocation },
  openAPIVersion: SupportedOpenAPIVersions
): IterableIterator<ShapePatch> {
  if (
    diff.kind !== ShapeDiffResultKind.UnmatchedType ||
    diff.keyword !== JsonSchemaKnownKeyword.type
  )
    return;

  const currentPropertySchema = jsonPointerHelpers.get(
    schema,
    diff.propertyPath
  );
  const alreadyOneOf = Array.isArray(currentPropertySchema.oneOf);

  function makeOneOfOperations() {
    const groupedOperations: OperationGroup[] = [];
    if (alreadyOneOf) {
      let baseSchema = Schema.baseFromValue(diff.example, openAPIVersion);

      if (
        !currentPropertySchema.oneOf.find((branchSchema) =>
          Schema.equals(baseSchema, branchSchema)
        )
      ) {
        groupedOperations.push(
          OperationGroup.create(`add new oneOf type to ${diff.key}`, {
            op: 'add',
            path: jsonPointerHelpers.append(diff.propertyPath, 'oneOf', '-'), // "-" indicates append to array
            value: baseSchema,
          })
        );
      }
    } else {
      let mergeOperations = [
        ...Schema.mergeOperations(Schema.clone(currentPropertySchema), {
          oneOf: [
            Schema.clone(currentPropertySchema),
            Schema.baseFromValue(diff.example, openAPIVersion),
          ],
        }),
      ];

      groupedOperations.push(
        OperationGroup.create(
          `replace ${diff.key} with a one of containing both types`,
          ...mergeOperations.map((op) => ({
            ...op,
            path: jsonPointerHelpers.join(diff.propertyPath, op.path),
          }))
        )
      );
    }

    return groupedOperations;
  }

  function changeTypeOperations() {
    return [
      OperationGroup.create(`change ${diff.key} type`, {
        op: 'replace',
        path: jsonPointerHelpers.append(diff.propertyPath),
        // handles removal of keys that are no longer allowed
        value: Schema.merge(
          currentPropertySchema,
          Schema.baseFromValue(diff.example, openAPIVersion)
        ),
      }),
    ];
  }

  if (openAPIVersion === '3.0.x') {
    if (diff.example === null) {
      yield {
        description: `make ${diff.key} null`,
        diff,
        impact: [
          PatchImpact.Addition,
          !shapeContext.location
            ? PatchImpact.BackwardsCompatibilityUnknown
            : 'inRequest' in shapeContext.location
            ? PatchImpact.BackwardsCompatible
            : PatchImpact.BackwardsIncompatible,
        ],
        groupedOperations: [
          OperationGroup.create(`make ${diff.key} nullable`, {
            op: 'replace',
            path: jsonPointerHelpers.append(diff.propertyPath, 'nullable'),
            // handles removal of keys that are no longer allowed
            value: true,
          }),
        ],
        shouldRegeneratePatches: false,
        interaction,
      };
    } else {
      // option one: convert to a one-off
      yield {
        description: `make ${diff.key} oneOf`,
        diff,
        impact: [
          PatchImpact.Addition,
          !shapeContext.location
            ? PatchImpact.BackwardsCompatibilityUnknown
            : 'inRequest' in shapeContext.location
            ? PatchImpact.BackwardsCompatible
            : PatchImpact.BackwardsIncompatible,
        ],
        groupedOperations: makeOneOfOperations(),
        shouldRegeneratePatches: true,
        interaction,
      };

      // option two: change the type
      yield {
        diff,
        description: `change type of ${diff.key}`,
        impact: [PatchImpact.BackwardsIncompatible],
        groupedOperations: changeTypeOperations(),
        shouldRegeneratePatches: false,
        interaction,
      };
    }
  } else if (openAPIVersion === '3.1.x') {
    if (diff.example === null) {
      const subschema = jsonPointerHelpers.get(schema, diff.propertyPath);
      const schemaType = Array.isArray(subschema.type)
        ? [...subschema.type, 'null']
        : subschema.type
        ? [subschema.type, 'null']
        : ['null'];

      yield {
        description: `make ${diff.key} null`,
        diff,
        impact: [
          PatchImpact.Addition,
          !shapeContext.location
            ? PatchImpact.BackwardsCompatibilityUnknown
            : 'inRequest' in shapeContext.location
            ? PatchImpact.BackwardsCompatible
            : PatchImpact.BackwardsIncompatible,
        ],
        groupedOperations: [
          OperationGroup.create(`make ${diff.key} null`, {
            op: 'replace',
            path: jsonPointerHelpers.append(diff.propertyPath, 'type'),
            // handles removal of keys that are no longer allowed
            value: schemaType,
          }),
        ],
        shouldRegeneratePatches: true,
        interaction,
      };
    } else {
      yield {
        description: `make ${diff.key} oneOf`,
        diff,
        impact: [
          PatchImpact.Addition,
          !shapeContext.location
            ? PatchImpact.BackwardsCompatibilityUnknown
            : 'inRequest' in shapeContext.location
            ? PatchImpact.BackwardsCompatible
            : PatchImpact.BackwardsIncompatible,
        ],
        groupedOperations: makeOneOfOperations(),
        shouldRegeneratePatches: true,
        interaction,
      };

      // option two: change the type
      yield {
        diff,
        description: `change type of ${diff.key}`,
        impact: [PatchImpact.BackwardsIncompatible],
        groupedOperations: changeTypeOperations(),
        shouldRegeneratePatches: false,
        interaction,
      };
    }
  }
}
