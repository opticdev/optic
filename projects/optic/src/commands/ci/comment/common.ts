import {
  Severity,
  getOperationsChangedLabel,
  getOperationsChanged,
} from '@useoptic/openapi-utilities';
import { CiRunDetails } from '../../../utils/ci-data';
import { GroupedDiffs } from '@useoptic/openapi-utilities/build/openapi3/group-diff';

const getChecksLabel = (
  results: CiRunDetails['completed'][number]['comparison']['results'],
  severity: Severity
) => {
  let totalChecks = results.length;
  let failingChecks = 0;
  let exemptedFailingChecks = 0;

  for (const result of results) {
    if (result.passed) continue;
    if (result.severity < severity) continue;
    if (result.exempted) exemptedFailingChecks += 1;
    else failingChecks += 1;
  }

  const exemptedChunk =
    exemptedFailingChecks > 0 ? `, ${exemptedFailingChecks} exempted` : '';

  return failingChecks > 0
    ? `‚ö†Ô∏è **${failingChecks}**/**${totalChecks}** failed${exemptedChunk}`
    : totalChecks > 0
    ? `‚úÖ **${totalChecks}** passed${exemptedChunk}`
    : `‚ÑπÔ∏è No automated checks have run`;
};

export const COMPARE_SUMMARY_IDENTIFIER = `optic-comment-3UsoJCz_Z0SpGLo5Vjw6o`;

function getOperationsText(
  groupedDiffs: GroupedDiffs,
  options: { webUrl?: string | null; verbose: boolean; labelJoiner?: string }
) {
  const ops = getOperationsChanged(groupedDiffs);

  const operationsText = options.verbose
    ? [
        ...[...ops.added].map((o) => `\`${o}\` (added)`),
        ...[...ops.changed].map((o) => `\`${o}\` (changed)`),
        ...[...ops.removed].map((o) => `\`${o}\` (removed)`),
      ].join('\n')
    : '';
  return `${getOperationsChangedLabel(groupedDiffs, {
    joiner: options.labelJoiner,
  })}

  ${operationsText}
`;
}

const getCaptureIssuesLabel = ({
  unmatchedInteractions,
  mismatchedEndpoints,
}: {
  unmatchedInteractions: number;
  mismatchedEndpoints: number;
}) => {
  return [
    ...(unmatchedInteractions
      ? [
          `üÜï ${unmatchedInteractions} undocumented path${
            unmatchedInteractions > 1 ? 's' : ''
          }`,
        ]
      : []),
    ...(mismatchedEndpoints
      ? [
          `‚ö†Ô∏è  ${mismatchedEndpoints} mismatch${
            mismatchedEndpoints > 1 ? 'es' : ''
          }`,
        ]
      : []),
  ].join('\n');
};

export const generateCompareSummaryMarkdown = (
  commit: { sha: string },
  results: CiRunDetails,
  options: { verbose: boolean }
) => {
  const anyCompletedHasWarning = results.completed.some(
    (s) => s.warnings.length > 0
  );
  const anyCompletedHasCapture = results.completed.some((s) => s.capture);
  return `
<!--
DO NOT MODIFY
app_id: ${COMPARE_SUMMARY_IDENTIFIER}
commit_sha: ${commit.sha}
-->
${
  results.completed.length > 0
    ? `### APIs Changed

<table>
<thead>
<tr>
<th>API</th>
<th>Changes</th>
<th>Rules</th>
${anyCompletedHasWarning ? '<th>Warnings</th>' : ''}
${anyCompletedHasCapture ? '<th>Tests</th>' : ''}
<th></th>
</tr>
</thead>
<tbody>
${results.completed
  .map(
    (s) =>
      `<tr>
<td>

${s.apiName}

</td>
<td>

${getOperationsText(s.comparison.groupedDiffs, {
  webUrl: s.opticWebUrl,
  verbose: options.verbose,
  labelJoiner: ',\n',
})}

</td>
<td>

${getChecksLabel(s.comparison.results, results.severity)}

</td>

${anyCompletedHasWarning ? `<td>${s.warnings.join('\n')}</td>` : ''}

${
  anyCompletedHasCapture
    ? `

<td>

${
  s.capture
    ? s.capture.success
      ? s.capture.mismatchedEndpoints || s.capture.unmatchedInteractions
        ? getCaptureIssuesLabel({
            unmatchedInteractions: s.capture.unmatchedInteractions,
            mismatchedEndpoints: s.capture.mismatchedEndpoints,
          })
        : `‚úÖ ${s.capture.percentCovered}% coverage`
      : '‚ùå¬†Failed to run'
    : ''
}

</td>

`
    : ''
}

<td>

${s.opticWebUrl ? `[View report](${s.opticWebUrl})` : ''}

</td>
</tr>`
  )
  .join('\n')}
</tbody>
</table>
`
    : ''
}
${
  results.failed.length > 0
    ? `### Errors running optic

<table>
<thead>
<tr>
<th>API</th>
<th>Error</th>
</tr>
</thead>
<tbody>
${results.failed
  .map(
    (s) => `<tr>
<td>${s.apiName}</td>
<td>

${'```'}
${s.error}
${'```'}

</td>
</tr>`
  )
  .join('\n')}
</tbody>
</table>
`
    : ''
}

Summary of API changes for commit (${commit.sha})

${
  results.noop.length > 0
    ? `${
        results.noop.length === 1 ? '1 API' : `${results.noop.length} APIs`
      } had no changes.`
    : ''
}`;
};

const LOADING_BANNER_IDENTIFIER = 'LOADING-BANNER-zRdRfihnV6eN5__wdKq5A';
export const getHasLoadingBanner = (body: string) =>
  new RegExp(LOADING_BANNER_IDENTIFIER).test(body);

export const createLoadingBanner = (commitSha: string) => `
<!-- ${LOADING_BANNER_IDENTIFIER} -->
> ‚ÑπÔ∏è Running update for commit ${commitSha}`;
