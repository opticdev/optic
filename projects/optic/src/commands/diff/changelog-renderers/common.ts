import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { FlatOpenAPIV3, FlatOpenAPIV3_1 } from '@useoptic/openapi-utilities';
import { Diff } from '@useoptic/openapi-utilities/build/openapi3/group-diff';

export type SpecInput = {
  from: FlatOpenAPIV3.Document | FlatOpenAPIV3_1.Document;
  to: FlatOpenAPIV3.Document | FlatOpenAPIV3_1.Document;
};

export function getRootBodyPath(path: string): string {
  const parts = jsonPointerHelpers.decode(path);
  if (parts[3] === 'responses') {
    return jsonPointerHelpers.compile(parts.slice(0, 7));
  } else {
    return jsonPointerHelpers.compile(parts.slice(0, 6));
  }
}
export function interpretFieldLevelDiffs(
  specs: {
    from: FlatOpenAPIV3.Document | FlatOpenAPIV3_1.Document;
    to: FlatOpenAPIV3.Document | FlatOpenAPIV3_1.Document;
  },
  diffs: Record<string, { diffs: Diff[] }>
): Diff[] {
  return Object.entries(diffs)
    .filter(([_, { diffs }]) => diffs.length > 0)
    .map(([key, { diffs }]) => {
      const firstDiffPath = diffs[0].after ?? diffs[0].before;

      // TODO figure out if we need to handle path reconciliation here
      // This might only be an issue with nested enums + rearranging + removing keys with oneOf/anyOf/allOf
      const absolutePath = jsonPointerHelpers.join(
        getRootBodyPath(firstDiffPath),
        key
      );
      const beforeRaw = jsonPointerHelpers.tryGet(specs.from, absolutePath);
      const afterRaw = jsonPointerHelpers.tryGet(specs.to, absolutePath);
      const before = beforeRaw.match ? absolutePath : undefined;
      const after = afterRaw.match ? absolutePath : undefined;

      return {
        before,
        after,
        trail: key,
        change: before && after ? 'changed' : before ? 'removed' : 'added',
      } as Diff;
    });
}
