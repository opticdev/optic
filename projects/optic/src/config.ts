import fs from 'node:fs/promises';
import fetch from 'node-fetch';
import yaml from 'js-yaml';
import { UserError, isTruthyStringValue } from '@useoptic/openapi-utilities';
import Ajv from 'ajv';
import os from 'os';
import path from 'node:path';
import { createOpticClient, OpticBackendClient } from './client';
import * as Git from './utils/git-utils';
import { logger } from './logger';
import { Static, Type } from '@sinclair/typebox';
import Handlebars from 'handlebars';
import * as dotenv from 'dotenv';
import { anonymizeOrgToken, anonymizeUserToken } from './client/optic-backend';

export enum VCS {
  Git = 'git',
  Cloud = 'cloud', // hosted in optic cloud
}

export const OPTIC_YML_NAME = 'optic.yml';
export const OPTIC_DEV_YML_NAME = 'optic.dev.yml';
export const USER_CONFIG_DIR = path.join(os.homedir(), '.config', 'optic');

export const USER_CONFIG_PATH =
  process.env.OPTIC_ENV === 'staging'
    ? path.join(USER_CONFIG_DIR, 'config.staging.json')
    : process.env.OPTIC_ENV === 'local'
      ? path.join(USER_CONFIG_DIR, 'config.local.json')
      : path.join(USER_CONFIG_DIR, 'config.json');

const checkIsInCi = (): boolean => {
  return process.env.CI !== undefined && isTruthyStringValue(process.env.CI);
};

const DefaultOpticCliConfig: OpticCliConfig = {
  isDefaultConfig: true,
  root: process.cwd(),
  configPath: undefined,
  ruleset: undefined,
  isAuthenticated: false,
  client: createOpticClient('no_token'),
  isInCi: checkIsInCi(),
};

const RequestSend = Type.Object({
  path: Type.String(),
  method: Type.Optional(
    Type.String({
      enum: ['GET', 'POST', 'PATCH', 'DELETE'],
    })
  ),
  data: Type.Optional(Type.Object({})),
  headers: Type.Optional(Type.Record(Type.String(), Type.String())),
});

const RequestRun = Type.Optional(
  Type.Object({
    command: Type.String(),
    proxy_variable: Type.Optional(Type.String()),
  })
);

const CaptureConfigData = Type.Object({
  config: Type.Optional(
    Type.Object({
      request_concurrency: Type.Optional(Type.Number()),
    })
  ),
  server: Type.Object({
    dir: Type.Optional(Type.String()),
    command: Type.Optional(Type.String()),
    url: Type.String(),
    ready_endpoint: Type.Optional(Type.String()),
    ready_interval: Type.Optional(Type.Number()),
    ready_timeout: Type.Optional(Type.Number()),
  }),
  requests: Type.Object({
    // one of these is technically required, but that's enforced at runtime
    run: Type.Optional(RequestRun),
    send: Type.Optional(Type.Array(RequestSend)),
  }),
});

export type CaptureConfigData = Static<typeof CaptureConfigData>;
export type ServerConfig = CaptureConfigData['server'];
export type RequestSend = Static<typeof RequestSend>;
export type RequestRun = Static<typeof RequestRun>;
export type CaptureConfigConfig = CaptureConfigData['config'];

export const ProjectYmlConfig = Type.Object({
  extends: Type.Optional(Type.String()),
  ruleset: Type.Optional(
    Type.Array(
      Type.Union([
        Type.String(),
        Type.Object(
          {},
          {
            minProperties: 1,
            maxProperties: 1,
          }
        ),
      ])
    )
  ),
  capture: Type.Optional(Type.Record(Type.String(), CaptureConfigData)),
  external_refs: Type.Optional(
    Type.Object({
      resolve_headers: Type.Optional(
        Type.Array(
          Type.Object({
            headers: Type.Record(Type.String(), Type.String()),
            url_prefix: Type.Optional(Type.String()),
          })
        )
      ),
    })
  ),
});
export type ProjectYmlConfig = Static<typeof ProjectYmlConfig>;
export type ConfigRuleset = { name: string; config: unknown };

export type OpticCliConfig = Omit<ProjectYmlConfig, 'ruleset'> & {
  isDefaultConfig: boolean;

  ruleset?: ConfigRuleset[];

  // path to the loaded config, or undefined if it was the default config
  configPath?: string;

  // undefined if git repo root was detected, otherwise the path to the git repo root
  root: string;

  // the detected vcs
  vcs?: {
    type: VCS;
    sha: string;
    // Set of absolute paths
    diffSet: Set<string>;
  };

  isAuthenticated: boolean;
  authenticationType?: 'user' | 'env';
  userId?: string;
  client: OpticBackendClient;

  isInCi: boolean;
};

const ajv = new Ajv();
const validateConfigSchema = ajv.compile(ProjectYmlConfig);

// attempt to find an optic.yml file, or return undefined if none can be found
// Starts from cwd, and looks up until the top level dir (usually the git root or cwd)
export async function detectCliConfig(
  topLevelDir: string
): Promise<string | undefined> {
  let currentDir = process.cwd();
  // search up from pwd to specified dir
  while (topLevelDir.length <= currentDir.length) {
    const expectedDevYmlPath = path.join(currentDir, OPTIC_DEV_YML_NAME);
    const expectedYmlPath = path.join(currentDir, OPTIC_YML_NAME);
    try {
      await fs.access(expectedYmlPath);
      return expectedYmlPath;
    } catch (e) {}

    try {
      await fs.access(expectedDevYmlPath);
      return expectedDevYmlPath;
    } catch (e) {}

    const next = path.dirname(currentDir);
    if (currentDir === next) break;
    currentDir = next;
  }

  return undefined;
}

export async function RenderTemplate(configPath: string): Promise<unknown> {
  const template = Handlebars.compile(await fs.readFile(configPath, 'utf-8'));
  const result = template(process.env);
  return yaml.load(result);
}

export async function loadCliConfig(
  configPath: string,
  client: OpticBackendClient
): Promise<OpticCliConfig> {
  const config = await RenderTemplate(configPath);

  validateConfig(config, configPath);
  await initializeRules(config as ProjectYmlConfig, client);

  const cliConfig = config as OpticCliConfig;
  cliConfig.root = path.dirname(configPath);
  cliConfig.configPath = configPath;

  return cliConfig;
}

export const validateConfig = (config: unknown, path: string) => {
  const result = validateConfigSchema(config);

  if (!result) {
    throw new UserError({
      message: `Configuration file \`${path}\` is invalid:\n${ajv.errorsText(
        validateConfigSchema.errors
      )}`,
    });
  }
};

export const initializeRules = async (
  config: ProjectYmlConfig,
  client: OpticBackendClient
) => {
  let rulesetMap: Map<string, ConfigRuleset> = new Map();
  let rawRulesets = config.ruleset ? config.ruleset : [];
  if (config.extends) {
    logger.debug(`Extending ruleset from ${config.extends}`);

    try {
      if (config.extends.startsWith('@')) {
        logger.error('Cloud rulesets are not supported');
      } else {
        // Assumption is that we're fetching a yaml file
        const response = await fetch(config.extends).then((response) => {
          if (response.status !== 200) {
            throw new Error(`received status code ${response.status}`);
          } else {
            return response.text();
          }
        });
        const parsed = yaml.load(response);
        rawRulesets.push(...(parsed as any).ruleset);
      }
    } catch (e) {
      console.error(e);
      console.log(
        `Failed to download the ruleset from ${config.extends}. Not using extended ruleset`
      );
    }
  }

  if (rawRulesets.length) {
    for (const ruleset of rawRulesets) {
      if (typeof ruleset === 'string') {
        rulesetMap.set(ruleset, { name: ruleset, config: {} });
      } else if (typeof ruleset === 'object' && ruleset !== null) {
        const keys = Object.keys(ruleset);
        if (keys.length !== 1) {
          throw new UserError({
            message: `Configuration error: empty ruleset configuration`,
          });
        } else {
          const name = keys[0];
          const config = ruleset[name] || {};
          rulesetMap.set(name, { name, config });
        }
      } else {
        throw new UserError({
          message: 'Configuration error: unexpected ruleset format',
        });
      }
    }

    config.ruleset = [...rulesetMap.values()];
  }
};

export type UserConfig = {
  token: string;
};

export async function readUserConfig(): Promise<UserConfig | null> {
  try {
    const validator = ajv.compile({
      type: 'object',
      properties: {
        token: {
          type: 'string',
        },
      },
      required: ['token'],
    });
    const unvalidatedConfig = JSON.parse(
      await fs.readFile(USER_CONFIG_PATH, 'utf-8')
    );
    const result = validator(unvalidatedConfig);
    if (!result) {
      return null;
    }

    return unvalidatedConfig as UserConfig;
  } catch (e) {
    return null;
  }
}

async function getYmlOrJsonChanges(gitRoot: string): Promise<Set<string>> {
  const status = await Git.gitStatus();
  const ymlOrJsonWithChanges = status
    .split('\n')
    .filter((line) => /\.(json|ya?ml)$/i.test(line))
    .map((line) => {
      const file = line.split(' ').slice(-1)[0].trim();
      return path.join(gitRoot, file);
    });

  return new Set(ymlOrJsonWithChanges);
}

export async function initializeConfig(): Promise<OpticCliConfig> {
  let cliConfig: OpticCliConfig = DefaultOpticCliConfig;
  const userConfig = await readUserConfig();
  const maybeEnvToken = process.env.OPTIC_TOKEN;
  const maybeUserToken = userConfig?.token
    ? Buffer.from(userConfig.token, 'base64').toString()
    : null;
  const token = maybeEnvToken || maybeUserToken;
  if (token) {
    cliConfig.authenticationType = maybeEnvToken
      ? 'env'
      : maybeUserToken
        ? 'user'
        : undefined;
    cliConfig.isAuthenticated = true;
    cliConfig.userId = token.startsWith('opat')
      ? anonymizeUserToken(token)
      : anonymizeOrgToken(token);
    cliConfig.client = createOpticClient(token);
  }

  if ((await Git.hasGit()) && (await Git.isInGitRepo())) {
    const gitRoot = await Git.getRootPath();
    const opticYmlPath = await detectCliConfig(gitRoot);

    if (opticYmlPath) {
      logger.debug(`Using config found at ${opticYmlPath}`);
      // if present, expect the .optic.env is in the same dir as the config file
      dotenv.config({
        path: path.join(path.dirname(opticYmlPath), '.optic.env'),
      });
      cliConfig = {
        ...cliConfig,
        ...(await loadCliConfig(opticYmlPath, cliConfig.client)),
        isDefaultConfig: false,
      };
    } else {
      // if present, expect the .optic.env in the current working dir
      dotenv.config({ path: path.join(process.cwd(), '.optic.env') });
      cliConfig.root = gitRoot;
    }

    try {
      cliConfig.vcs = {
        type: VCS.Git,
        sha: await Git.resolveGitRef('HEAD'),
        diffSet: await getYmlOrJsonChanges(gitRoot),
      };
    } catch (e) {
      // Git command can fail in a repo with no commits, we should treat this as having no commits
    }
  } else {
    // if present, expect the .optic.env in the current working dir
    const currentDir = process.cwd();
    dotenv.config({ path: path.join(currentDir, '.optic.env') });
    const opticYmlPath = await detectCliConfig(currentDir);

    if (opticYmlPath) {
      logger.debug(`Using config found at ${opticYmlPath}`);
      cliConfig = {
        ...cliConfig,
        ...(await loadCliConfig(opticYmlPath, cliConfig.client)),
        isDefaultConfig: false,
      };
    }
  }

  logger.debug(cliConfig);

  return cliConfig;
}
