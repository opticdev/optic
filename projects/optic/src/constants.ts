export const OPTIC_STANDARD_KEY = 'x-optic-standard';
export const OPTIC_URL_KEY = 'x-optic-url';
