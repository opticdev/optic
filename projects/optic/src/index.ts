#!/usr/bin/env node
import { initCli } from './init';

(async () => {
  const cli = await initCli();

  cli.parse(process.argv);
})();
