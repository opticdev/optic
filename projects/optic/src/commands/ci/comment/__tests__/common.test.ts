import { describe, test, expect } from '@jest/globals';
import { generateCompareSummaryMarkdown } from '../common';
import type { CiRunDetails } from '../../../../utils/ci-data';
import {
  defaultEmptySpec,
  diff,
  groupDiffsByEndpoint,
} from '@useoptic/openapi-utilities';
import { FlatOpenAPIV3 } from '@useoptic/openapi-utilities/src';

const from: FlatOpenAPIV3.Document = defaultEmptySpec as FlatOpenAPIV3.Document;
const to: FlatOpenAPIV3.Document = {
  ...defaultEmptySpec,
  paths: {
    '/api': {
      get: {
        responses: {
          '200': {
            description: 'hello',
          },
        },
      },
      post: {
        responses: {
          '200': {
            description: 'hello',
          },
        },
      },
    },
  },
} as FlatOpenAPIV3.Document;

const input: CiRunDetails = {
  completed: [
    {
      apiName: 'completed',
      opticWebUrl: 'https://app.useoptic.com/changelog',
      specUrl: 'https://app.useoptic.com/docs',
      warnings: [],
      comparison: {
        groupedDiffs: groupDiffsByEndpoint(
          {
            from,
            to,
          },
          diff(from, to),
          []
        ),
        results: [],
      },
    },
  ],
  failed: [
    {
      apiName: 'failed',
      error: 'blahblahblah',
    },
  ],
  noop: [
    {
      apiName: 'noop',
    },
  ],
  severity: 2,
};
describe('generateCompareSummaryMarkdown', () => {
  test('generates md output for passed, failed and noop', () => {
    expect(
      generateCompareSummaryMarkdown({ sha: '123' }, input, {
        verbose: false,
      })
    ).toMatchSnapshot();
  });

  test('generates md output for passed, failed and noop for verbose', () => {
    expect(
      generateCompareSummaryMarkdown({ sha: '123' }, input, {
        verbose: true,
      })
    ).toMatchSnapshot();
  });

  test('only passed', () => {
    expect(
      generateCompareSummaryMarkdown(
        { sha: '123' },
        {
          completed: input.completed,
          failed: [],
          noop: [],
          severity: 2,
        },
        {
          verbose: false,
        }
      )
    ).toMatchSnapshot();
  });

  test('only failed', () => {
    expect(
      generateCompareSummaryMarkdown(
        { sha: '123' },
        {
          completed: [],
          failed: input.failed,
          noop: [],
          severity: 2,
        },
        {
          verbose: false,
        }
      )
    ).toMatchSnapshot();
  });

  test('only noop', () => {
    expect(
      generateCompareSummaryMarkdown(
        { sha: '123' },
        {
          completed: [],
          failed: [],
          noop: input.noop,
          severity: 2,
        },
        {
          verbose: false,
        }
      )
    ).toMatchSnapshot();
  });

  test('when run without optic cloud', () => {
    expect(
      generateCompareSummaryMarkdown(
        { sha: '123' },
        {
          completed: [
            {
              apiName: 'no-optic-cloud',
              warnings: [],
              comparison: {
                groupedDiffs: groupDiffsByEndpoint(
                  {
                    from,
                    to,
                  },
                  diff(from, to),
                  []
                ),
                results: [],
              },
            },
          ],
          failed: [],
          noop: [],
          severity: 2,
        },
        {
          verbose: false,
        }
      )
    ).toMatchSnapshot();
  });
});
