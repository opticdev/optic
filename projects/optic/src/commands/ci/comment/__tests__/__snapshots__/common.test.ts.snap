// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateCompareSummaryMarkdown generates md output for passed, failed and noop 1`] = `
"
<!--
DO NOT MODIFY
app_id: optic-comment-3UsoJCz_Z0SpGLo5Vjw6o
commit_sha: 123
-->
### APIs Changed

<table>
<thead>
<tr>
<th>API</th>
<th>Changes</th>
<th>Rules</th>


<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>

completed

</td>
<td>

2 operations added

  


</td>
<td>

ℹ️ No automated checks have run

</td>





<td>

[View report](https://app.useoptic.com/changelog)

</td>
</tr>
</tbody>
</table>

### Errors running optic

<table>
<thead>
<tr>
<th>API</th>
<th>Error</th>
</tr>
</thead>
<tbody>
<tr>
<td>failed</td>
<td>

\`\`\`
blahblahblah
\`\`\`

</td>
</tr>
</tbody>
</table>


Summary of API changes for commit (123)

1 API had no changes."
`;

exports[`generateCompareSummaryMarkdown generates md output for passed, failed and noop for verbose 1`] = `
"
<!--
DO NOT MODIFY
app_id: optic-comment-3UsoJCz_Z0SpGLo5Vjw6o
commit_sha: 123
-->
### APIs Changed

<table>
<thead>
<tr>
<th>API</th>
<th>Changes</th>
<th>Rules</th>


<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>

completed

</td>
<td>

2 operations added

  \`GET /api\` (added)
\`POST /api\` (added)


</td>
<td>

ℹ️ No automated checks have run

</td>





<td>

[View report](https://app.useoptic.com/changelog)

</td>
</tr>
</tbody>
</table>

### Errors running optic

<table>
<thead>
<tr>
<th>API</th>
<th>Error</th>
</tr>
</thead>
<tbody>
<tr>
<td>failed</td>
<td>

\`\`\`
blahblahblah
\`\`\`

</td>
</tr>
</tbody>
</table>


Summary of API changes for commit (123)

1 API had no changes."
`;

exports[`generateCompareSummaryMarkdown only failed 1`] = `
"
<!--
DO NOT MODIFY
app_id: optic-comment-3UsoJCz_Z0SpGLo5Vjw6o
commit_sha: 123
-->

### Errors running optic

<table>
<thead>
<tr>
<th>API</th>
<th>Error</th>
</tr>
</thead>
<tbody>
<tr>
<td>failed</td>
<td>

\`\`\`
blahblahblah
\`\`\`

</td>
</tr>
</tbody>
</table>


Summary of API changes for commit (123)

"
`;

exports[`generateCompareSummaryMarkdown only noop 1`] = `
"
<!--
DO NOT MODIFY
app_id: optic-comment-3UsoJCz_Z0SpGLo5Vjw6o
commit_sha: 123
-->



Summary of API changes for commit (123)

1 API had no changes."
`;

exports[`generateCompareSummaryMarkdown only passed 1`] = `
"
<!--
DO NOT MODIFY
app_id: optic-comment-3UsoJCz_Z0SpGLo5Vjw6o
commit_sha: 123
-->
### APIs Changed

<table>
<thead>
<tr>
<th>API</th>
<th>Changes</th>
<th>Rules</th>


<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>

completed

</td>
<td>

2 operations added

  


</td>
<td>

ℹ️ No automated checks have run

</td>





<td>

[View report](https://app.useoptic.com/changelog)

</td>
</tr>
</tbody>
</table>



Summary of API changes for commit (123)

"
`;

exports[`generateCompareSummaryMarkdown when run without optic cloud 1`] = `
"
<!--
DO NOT MODIFY
app_id: optic-comment-3UsoJCz_Z0SpGLo5Vjw6o
commit_sha: 123
-->
### APIs Changed

<table>
<thead>
<tr>
<th>API</th>
<th>Changes</th>
<th>Rules</th>


<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>

no-optic-cloud

</td>
<td>

2 operations added

  


</td>
<td>

ℹ️ No automated checks have run

</td>





<td>



</td>
</tr>
</tbody>
</table>



Summary of API changes for commit (123)

"
`;
